src-tauri/src/note.rs

```diff
@@ -13,26 +14,97 @@ impl Note {
     }
 
+    /// Reads the contents of the given note.
+    pub fn read(&self, name : &str) -> NoteResult<String> {
+      let readme = Path::new(&self.get_note_path(name)).join("README.md");
+      Ok(fs::read_to_string(readme)?)
+    }
+
+    /// Creates a new note and return it's name.
+    pub fn create(&self) -> NoteResult<String> {
+      let mut name = String::from("Untitled");
+      let mut path = self.get_note_path(&name);
+      let mut n = 0;
+      while path.as_path().exists() {
+        n += 1;
+        name = format!("Untitled {}", n);
+        path = self.get_note_path(&name);
+      }
+
+      fs::create_dir(path.clone())?;
+      let readme = Path::new(&path).join("README.md");
+      fs::write(readme, format!("# {}\n", name))?;
+      let tags = Path::new(&path).join("tags.txt");
+      fs::write(tags, "")?;
+
+      Ok(name)
+    }
+
+    /// Renames a note.
+    pub fn rename(&self, old_name: &str, new_name: &str) -> NoteResult<()> {
+      let old_path = self.get_note_path(old_name);
+      let new_path = self.get_note_path(new_name);
+      Ok(fs::rename(old_path, new_path)?)
+    }
+
+    /// Writes the content of the given notes.
+    pub fn write(&self, name: &str, content: &str) -> NoteResult<()> {
+      let readme = Path::new(&self.get_note_path(name)).join("README.md");
+      fs::write(readme, content.as_bytes())?;
+      Ok(())
+    }
+
+    /// Removes the given note.
+    pub fn remove(&self, name: &str) -> NoteResult<()> {
+      let path = self.get_note_path(name);
+      Ok(fs::remove_dir_all(path)?)
+    }
+
+    /// Reads the tags associated with the given note.
+    pub fn read_tags(&self, name: &str) -> NoteResult<Vec<String>> {
+      let tags = Path::new(&self.get_note_path(name)).join("tags.txt");
+      let content = fs::read_to_string(tags)?;
+      let tags: Vec<String> = content
+        .lines()
+        .filter(|s| !s.is_empty())
+        .map(String::from)
+        .collect();
+      Ok (tags)
+    }
+
+    /// Replaces the tags of tags of the given note.
+    pub fn write_tags(&self, name: &str, tags: &Vec<String>) -> NoteResult<()> {
+      let content = tags.join("\n");
+      let tags = Path::new(&self.get_note_path(name)).join("tags.txt");
+      fs::write(tags, content.as_bytes())?;
+      Ok(())
+    }
+
+    fn get_note_path(&self, name: &str) -> PathBuf {
+      let mut path = self.config.get_base_path();
+      path.push(name);
+      path
+    }
 }
 
 /// list all directories in `base_path` that have a `README.md` file inside
-fn get_note_names(base_path: PathBuf) -> Result<Vec<String>> {
+fn get_note_names(base_path: PathBuf) -> NoteResult<Vec<String>> {
   let mut note_names = Vec::<String>::new();
   let dir_entries = std::fs::read_dir(base_path)?;
   for dir_entry in dir_entries {
-    match get_note_name(dir_entry) {
+    match get_note_name(dir_entry?) {
       Ok(note_name) => note_names.push(note_name),
       Err(e) => error_handling(e.to_string())
     }
```

Comment:
This should not break on Err but continue in my opinion. Currently, if `get_note_name` returns an error, the for loop will break and no more directories will be searched.

Reply:
In my opinion, it is quite weird, that Rust wraps each directory entry in a Result. There is some information about this at [read_dir documentation](https://doc.rust-lang.org/std/fs/fn.read_dir.html):

> The iterator will yield instances of [io::Result](https://doc.rust-lang.org/std/io/type.Result.html)<[DirEntry](https://doc.rust-lang.org/std/fs/struct.DirEntry.html)>. New errors may be encountered after an iterator is initially constructed.

Therefore, a `dir_entry` result becomes erroneous, when there is an error traversing the directory. Since we also let the method fail when `read_dir` fails two lines above an we can assume that all following Results are erroneous, I think we can keep this.